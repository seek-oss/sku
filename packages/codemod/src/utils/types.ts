export type Transform = (source: string) => Promise<string | null>;

export const isValidTransform = (fn: unknown): fn is Transform =>
  typeof fn === 'function';
