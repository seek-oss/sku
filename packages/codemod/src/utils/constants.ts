type Codemod = {
  description: string;
  value: string;
};

export const CODEMODS = [
  {
    description:
      'Transform all webpack loadable imports to Vite loadable imports',
    value: 'transform-vite-loadable',
  },
  {
    description: 'Convert Jest tests to Vitest',
    value: 'jest-to-vitest',
  },
  /* [add-sku-codemod-generator: codemod] */
] as const satisfies Codemod[];

export type CodemodName = (typeof CODEMODS)[number]['value'];
