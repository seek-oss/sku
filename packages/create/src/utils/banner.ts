import chalk from 'chalk';

export const banner = (
  type: 'info' | 'success' | 'warning' | 'error',
  title: string,
  messages: string[] = [],
) => {
  const colorMap = {
    info: chalk.blue,
    success: chalk.green,
    warning: chalk.yellow,
    error: chalk.red,
  };

  const color = colorMap[type];
  let icon = 'ℹ️';
  if (type === 'success') {
    icon = '✅';
  } else if (type === 'error') {
    icon = '❌';
  }

  console.log();
  console.log(color(`${icon} ${title}`));

  if (messages.length > 0) {
    console.log();
    messages.forEach((message) => {
      console.log(`  ${message}`);
    });
  }

  console.log();
};
