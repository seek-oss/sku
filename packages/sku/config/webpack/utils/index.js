const loaders = require('./loaders');
const { resolvePackage } = require('./resolvePackage');

module.exports = {
  ...loaders,
  resolvePackage,
  TYPESCRIPT: /\.(ts|tsx)$/,
  JAVASCRIPT: /\.js$/,
  LESS: /\.less$/,
  IMAGE: [/\.bmp$/, /\.gif$/, /\.jpe?g$/, /\.png$/],
  SVG: /\.svg$/,
};
