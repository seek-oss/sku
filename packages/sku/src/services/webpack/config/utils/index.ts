import { extensions } from '../../../../config/fileResolutionExtensions.js';

export * from './loaders.js';
export { resolvePackage } from './resolvePackage.js';

export const TYPESCRIPT = new RegExp(`\.(${extensions.ts.join('|')})$`);
export const JAVASCRIPT = new RegExp(`\.(${extensions.js.join('|')})$`);
export const IMAGE = [
  /\.bmp$/,
  /\.gif$/,
  /\.jpe?g$/,
  /\.png$/,
  /\.webp$/,
  /\.avif$/,
];
export const SVG = /\.svg$/;
