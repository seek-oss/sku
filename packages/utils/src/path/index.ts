export { toPosixPath } from './toPosixPath.js';
