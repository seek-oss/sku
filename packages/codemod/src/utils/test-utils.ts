import { readFile } from 'node:fs/promises';
import path from 'node:path';
import { expect } from 'vitest';
import { isValidTransform, type Transform } from './types.js';

type FileInfo = {
  source: string;
  path: string;
};

type InlineTestProps = {
  transform: Transform;
  input: FileInfo;
  expectedOutput: string;
};

type TestProps = {
  dirName: string;
  transformName: string;
  fixtureName?: string;
  testOptions?: { extension?: string };
};

const runInlineTest = async ({
  transform,
  input,
  expectedOutput,
}: InlineTestProps) => {
  if (!isValidTransform(transform)) {
    throw new Error('Invalid transform function provided');
  }

  const output = transform(input.source);
  if (!output) {
    throw new Error('Transform failed');
  }

  expect(output.trim()).toEqual(expectedOutput.trim());

  return output;
};

export const runTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const testFixtureName = fixtureName || transformName;

  const fixtureDir = path.join(dirName, '..', '__testfixtures__');
  const inputPath = path.join(
    fixtureDir,
    `${testFixtureName}.input.${testOptions.extension}`,
  );

  // Assumes transform is one level up from __tests__ directory
  const transformImport = import(path.join(dirName, '..', transformName));
  const sourceRead = readFile(inputPath, 'utf8');
  const expectedOutputRead = readFile(
    path.join(fixtureDir, `${testFixtureName}.output.${testOptions.extension}`),
    'utf8',
  );

  const [{ transform }, source, expectedOutput] = await Promise.all([
    transformImport,
    sourceRead,
    expectedOutputRead,
  ]);

  return runInlineTest({
    transform,
    input: {
      path: inputPath,
      source,
    },
    expectedOutput,
  });
};

export const runNoChangeTest = async ({
  dirName,
  transformName,
  fixtureName,
  testOptions = {},
}: TestProps) => {
  const inputPath = path.join(
    dirName,
    '..',
    '__testfixtures__',
    `${fixtureName}.unchanged.${testOptions.extension}`,
  );

  const sourceRead = readFile(inputPath, 'utf8');
  const transformImport = import(path.join(dirName, '..', transformName));

  const [{ transform }, source] = await Promise.all([
    transformImport,
    sourceRead,
  ]);
  const output = await transform(source);

  expect(output).toEqual(null);
};
