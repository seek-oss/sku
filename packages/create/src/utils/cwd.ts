import { join } from 'node:path';

let cwd = process.cwd();

export const setCwd = (dir: string) => {
  cwd = dir;
};

export const getCwd = () => cwd;

export const getPathFromCwd = (filePath: string) => join(cwd, filePath);
