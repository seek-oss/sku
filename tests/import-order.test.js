const path = require('path');
const fs = require('fs-extra');
const dedent = require('dedent');
const runSkuScriptInDir = require('../test/utils/runSkuScriptInDir');

const appDirectory = path.dirname(
  require.resolve('@fixtures/import-order/sku.config.js'),
);
const srcDirectory = path.join(appDirectory, 'src');
const testFile = (fileName) => path.join(srcDirectory, fileName);

expect.addSnapshotSerializer({
  serialize: (val) => val,
  test: (val) => typeof val === 'string',
});

const files = {
  'index.ts': dedent/* ts */ `
    import './reset'; // side-effect imports should stay put

    import LocalComponent from './LocalComponent'; // sibling
    import Parent from '../parent'; // parents
    import someModule from 'some-module'; // external
    import vanillaStyles from './vanillaStyles.css'; // styles
    import distantParent from '../../../parent'; // parents
    import myself from '.'; // index
    import path from 'path'; //  built-in
    import styles from './styles.less'; // styles
    import utils from 'src/utils'; // internal
  `,
  'relatives.js': dedent/* js */ `
    import aTreat from './a.treat';
    import bTreat from './b.treat';
    import cTreat from '../c.treat';

    import b from './b';
    import a from './a';
    import c from '../c';
  `,
  'vanilla.ts': dedent/* ts */ `
    import aStyle from './a.css';
    import bStyle from './b.css';
    import cStyle from '../c.css';

    import b from './b';
    import a from './a';
    import c from '../c';
  `,
};

describe('import order', () => {
  beforeAll(async () => {
    await fs.ensureDir(srcDirectory);
  });

  afterAll(async () => {
    await fs.remove(srcDirectory);
  });

  test.each(Object.keys(files))('imports are ordered: %s', async (fileName) => {
    const filePath = testFile(fileName);
    await fs.writeFile(filePath, files[fileName]);

    await runSkuScriptInDir('format', appDirectory);

    const result = await fs.readFile(filePath, { encoding: 'utf-8' });

    expect(result).toMatchSnapshot();
  });
});
