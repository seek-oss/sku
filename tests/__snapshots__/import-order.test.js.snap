// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import order imports are ordered: index.ts 1`] = `
import './reset'; // side-effect imports should stay put

import path from 'path'; //  built-in

import someModule from 'some-module'; // external

import utils from 'src/utils'; // internal

import distantParent from '../../../parent'; // parents
import Parent from '../parent'; // parents

import LocalComponent from './LocalComponent'; // sibling

import myself from '.'; // index

import styles from './styles.less'; // styles
import vanillaStyles from './vanillaStyles.css'; // styles

`;

exports[`import order imports are ordered: relatives.js 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aTreat from './a.treat';
import bTreat from './b.treat';
import cTreat from '../c.treat';

`;

exports[`import order imports are ordered: vanilla.ts 1`] = `
import c from '../c';

import a from './a';
import b from './b';

import aStyle from './a.css';
import bStyle from './b.css';
import cStyle from '../c.css';

`;
